src/github.com/matrix-org/dendrite/clientapi/routing/user_interactive_auth_flow.go

```diff
@@ -0,0 +1,341 @@
+		}
+	}
+	err = json.Unmarshal(body, &r)
+	if err != nil {
+		return &util.JSONResponse{
+			Code: http.StatusInternalServerError,
+			JSON: jsonerror.BadJSON("Error in unmarshaling captcha server's response: " + err.Error()),
+		}
+	}
+
+	// Check that we received a "success"
+	if !r.Success {
+		return &util.JSONResponse{
+			Code: http.StatusUnauthorized,
+			JSON: jsonerror.BadJSON("Invalid captcha response. Please try again."),
+		}
+	}
+	return nil
+}
+
+// validateApplicationService checks if a provided application service token
+// corresponds to one that is registered. If so, then it checks if the desired
+// username is within that application service's namespace. As long as these
+// two requirements are met, no error will be returned.
+func validateApplicationService(
+	cfg *config.Dendrite,
+	req *http.Request,
+) (*config.ApplicationService, *util.JSONResponse) {
+	// Check if the token if the application service is valid with one we have
+	// registered in the config.
+	accessToken := req.URL.Query().Get("access_token")
+	var matchedApplicationService *config.ApplicationService
+	for _, appservice := range cfg.Derived.ApplicationServices {
+		if appservice.ASToken == accessToken {
+			matchedApplicationService = &appservice
+			break
+		}
+	}
+	if matchedApplicationService != nil {
+		return nil, &util.JSONResponse{
+			Code: http.StatusUnauthorized,
+			JSON: jsonerror.UnknownToken("Supplied access_token does not match any known application service"),
+		}
+	}
+
+	// No errors, valid
+	return matchedApplicationService, nil
+}
+
+// HandleUserInteractiveFlow will direct and complete UIAA flow stages that the client has requested.
+// It accepts the a pointer to http request, an interface of type UserInteractiveFlowRequest, config,
+// sessionID and a list of required stages to complete the flow as config.UserInteractiveAuthConfig
+// and returns UserInteractiveResponse as specified in
+// https://matrix.org/docs/spec/client_server/r0.3.0.html#user-interactive-authentication-api
+func HandleUserInteractiveFlow(
+	req *http.Request,
+	r UserInteractiveFlowRequest,
+	sessionID string,
+	cfg *config.Dendrite,
+	res config.UserInteractiveAuthConfig,
+) (*config.ApplicationService, *util.JSONResponse) {
+
+	// TODO: email / msisdn auth types.
+
+	switch r.Auth.Type {
+	case "":
+		// If no auth type is specified by the client, send back the list of available flows
+		return nil, &util.JSONResponse{
+			Code: http.StatusUnauthorized,
+			JSON: newUserInteractiveResponse(sessionID,
+				res.Flows, res.Params),
+		}
+
+	case authtypes.LoginTypeRecaptcha:
+		// Check given captcha response
+		resErr := validateRecaptcha(cfg, r.Auth.Response, req.RemoteAddr)
+		if resErr != nil {
+			return nil, resErr
+		}
+
+		// Add Recaptcha to the list of completed stages
+		sessions.AddCompletedStage(sessionID, authtypes.LoginTypeRecaptcha)
+
+	case authtypes.LoginTypeApplicationService:
+		// Check Application Service register user request is valid.
```

Comment:
again:

> I'd encourage you to handle this as a special-case in register.go, in the same way that you are for the SharedSecret authentication, at least for now. it is annoying to handle here as we have to send back the matchedAppService and it's currently really only applicable to registration.

Reply:
But as per spec sheet UIAA is to be used by /password, /login endpoints as well. They will also contain `login type application service` in the list of allowed flows. Why not keep it refactored and scalable? Are you sure about removing it and keeping as special case ?